### PR TITLE
Add test coverage support to unit tests

### PR DESCRIPTION
**Pull request by GomathiselviS**
This fix collects and reports the test coverage results for unit tests.

### CODE REVIEW

File: roles/ansible-test/tasks/main.yaml [hunk 1]
```diff
@@ -28,9 +28,17 @@
...
   when: ansible_test_split_in > 1
 
+- name: Erase coverage report
+  args:
+    chdir: "{{ _test_location }}"
+    executable: /bin/bash
+  environment: "{{ ansible_test_environment | default({}) }}"
+  shell: "source {{ ansible_test_venv_path }}/bin/activate; {{ ansible_coverage_erase }}"
+  when: ansible_test_test_command == 'units'
+
```
Comment (pabelanger) [inline roles/ansible-test/tasks/main.yaml:38]: lets move this into a post-run playbook for units, eg:  playbooks/ansible-test-units-base/post.yaml

then add the playbook below to the ansible-test-units-base job

File: zuul.d/jobs.yaml [hunk 2]
```diff
@@ -894,6 +895,7 @@
     vars:
       ansible_test_python: 3.8
       ansible_test_command: units
+      ansible_coverage_report: "ansible-test coverage report"
```
Comment (pabelanger) [inline zuul.d/jobs.yaml:898]: lets move this into playbooks, I'd prefer not to have cmd syntax in jobs directly

File: roles/ansible-test/defaults/main.yaml [hunk 3]
```diff
@@ -14,3 +14,4 @@ ansible_test_collection_dir: ~/.ansible/collections/ansible_collections
 ansible_test_enable_ara: false
 ansible_test_split_in: 1
 ansible_test_do_number: 1
+ansible_coverage_report: ""
```
Comment (pabelanger) [inline roles/ansible-test/defaults/main.yaml:17]: lets default this in playbooks to empty, so we don't have to update all jobs for it

File: roles/ansible-test/tasks/main.yaml [hunk 4]
```diff
@@ -33,4 +33,13 @@
     chdir: "{{ _test_location }}"
     executable: /bin/bash
   environment: "{{ ansible_test_environment | default({}) }}"
-  shell: "source {{ ansible_test_venv_path }}/bin/activate; {{ ansible_test_executable }} {{ ansible_test_test_command }} {{ ansible_test_options }} --python {{ ansible_test_python }} -vvvv {{ _integration_targets }}"
+  shell: "source {{ ansible_test_venv_path }}/bin/activate; {{ ansible_test_executable }} {{ ansible_test_test_command }} {{ ansible_test_options }} --python {{ ansible_test_python }} -vvvv {{ _integration_targets }}; {{ ansible_coverage_report }}"
```
Comment (pabelanger) [inline roles/ansible-test/tasks/main.yaml:36]: I feel the generate coverage report will have to move into a post-run playbook, that is more inline with other log collection jobs.  We may have to hard code paths a little, to work around the ansible role issue you mentioned before.

File: roles/ansible-test/tasks/main.yaml [hunk 5]
```diff
@@ -33,4 +33,13 @@
...
+
+- name: Ensure controller directory exists
```
Comment (pabelanger) [inline roles/ansible-test/tasks/main.yaml:38]: same, I would like to see us create a new post-run playbook for our unit test jobs

File: playbooks/ansible-test-base/post.yaml [hunk 6]
```diff
@@ -1,10 +1,20 @@
 ---
 - hosts: controller
   tasks:
+    - name: Get collection namespace
```
Comment (pabelanger) [inline playbooks/ansible-test-base/post.yaml:4]: for this, we don't want to modify ansible-test-base/post.yaml, as this is limited to units.

Lets, create ansible-test-units-base/post.yaml not to affect other jobs.

**Comment by GomathiselviS** [comment]
recheck

**Comment by GomathiselviS** [comment]
recheck

File: roles/ansible-test/tasks/init_test_options.yaml [hunk 7]
```diff
@@ -24,7 +24,8 @@
     ansible_test_options: "{{ ansible_test_options }} --requirements"
   when: not ansible_test_collections
 
-- name: Disable options for unit tests
+- name: Enable coverage commands for unit tests
   set_fact:
-    ansible_test_options: ""
+    ansible_test_options: "--coverage"
+    ansible_coverage_report: "ansible-test coverage html"
```
Comment (pabelanger) [inline roles/ansible-test/tasks/init_test_options.yaml:30]: I'd like to see us move this command out of run into post-run.

File: roles/ansible-test/tasks/main.yaml (hunk 4 above)
Comment (pabelanger) [reply roles/ansible-test/tasks/main.yaml:36]: Just to confirm, this should not need to be done, if we move ansible_coverage_report to post-run phase.

File: roles/ansible-test/tasks/main.yaml [hunk 8]
```diff
@@ -33,4 +33,14 @@
...
+  file:
+    path: "{{ ansible_user_dir }}/zuul-output/logs/controller/coverage"
+    state: directory
+  when: ansible_test_test_command == 'units'
+
+- name: Create a symbolic link to the coverage directory
+  shell: "ln -s /home/zuul/.ansible/collections/ansible_collections/openvswitch/{{ _collection_namespace.stdout }}/tests/output/reports/coverage/index.html {{ ansible_user_dir }}/zuul-output/logs/controller/coverage/"
```
Comment (pabelanger) [inline roles/ansible-test/tasks/main.yaml:45]: ln -s {{ _test_location }}/tests/output ~/zuul-output/logs/controller/

File: playbooks/ansible-test-units-base/post.yaml [hunk 9]
```diff
@@ -1,8 +1,12 @@
 ---
 - hosts: controller
   tasks:
+    # - name: Copy report file
+    # shell: "cp /home/zuul/.ansible/collections/ansible_collections/openvswitch/{{ _collection_namespace.stdout }}/tests/output/reports/coverage/index.html {{ ansible_user_dir }}/zuul-output/logs/controller/coverage/"
+
     # TODO: Migrate to fetch-zuul-logs when
     # https://review.openstack.org/#/c/583346/ is merged.
+
     - name: Collect log output
```
Comment (pabelanger) [inline playbooks/ansible-test-units-base/post.yaml:10]: For here, try setting https://docs.ansible.com/ansible/latest/modules/synchronize_module.html#parameter-copy_links

So we copy symlinks, not the symlinks themself

File: playbooks/ansible-test-units-base/post.yaml [hunk 10]
```diff
@@ -1,12 +1,52 @@
...
+
+    - name: Create a symbolic link to the coverage directory
+      shell: "ln -s {{ test_location }}/tests/output {{ ansible_user_dir }}; cp -rf {{ ansible_user_dir}}/output {{ ansible_user_dir }}/zuul-output/logs/controller/"
+
+
     # TODO: Migrate to fetch-zuul-logs when
     # https://review.openstack.org/#/c/583346/ is merged.
+
     - name: Collect log output
-      no_log: true
```
Comment (pabelanger) [inline playbooks/ansible-test-units-base/post.yaml:7]: lets add this back, if working properly

File: roles/ansible-test/defaults/main.yaml (hunk 3 above)
Comment (pabelanger) [inline roles/ansible-test/defaults/main.yaml:17]: this doesn't look to be used, and can be removed

File: playbooks/ansible-test-units-base/post.yaml [hunk 11]
```diff
@@ -1,12 +1,53 @@
...
+        executable: /bin/bash
+      environment: "{{ ansible_test_environment | default({}) }}"
+      shell: "source {{ ansible_test_venv_path }}/bin/activate; {{ ansible_coverage_report }}"
+
+    - name: Ensure controller directory exists
+      file:
+        path: "{{ ansible_user_dir }}/zuul-output/logs/controller"
+        state: directory
+
+    - name: Create a symbolic link to the coverage directory
```
Comment (pabelanger) [inline playbooks/ansible-test-units-base/post.yaml:39]: k, we don't actually need to do a symlink here, I was suggesting we keep that in run.yaml, for now let just do

```
cp -a {{ test_location }}/tests/output/reports/coverage {{ ansible_user_dir }}/zuul-output/logs/controller/
```

File: playbooks/ansible-test-units-base/post.yaml [hunk 12]
```diff
@@ -1,12 +1,53 @@
...
+      shell: "cp -rfa {{ test_location }}/tests/output/reports/coverage {{ ansible_user_dir }}/zuul-output/logs/controller/"
+
+
     # TODO: Migrate to fetch-zuul-logs when
     # https://review.openstack.org/#/c/583346/ is merged.
+
     - name: Collect log output
-      no_log: true
       synchronize:
+        no_log: true
```
Comment (pabelanger) [inline playbooks/ansible-test-units-base/post.yaml:48]: wrong location, needs to be one level up

File: playbooks/ansible-test-units-base/post.yaml [hunk 13]
```diff
@@ -1,12 +1,53 @@
...
         dest: "{{ zuul.executor.log_root }}/"
         mode: pull
         src: "{{ ansible_user_dir }}/zuul-output/logs/"
         verify_host: true
+        copy_links: true
```
Comment (pabelanger) [inline playbooks/ansible-test-units-base/post.yaml:53]: no longer needed, lets also drop

**Pull request merged**